nixos/nix-daemon: Organize buildMachine options with a submodule

Valodim opened the pull request:
This PR picks up #47636. It changes the `nix.buildMachine` option from an attribute set to a submodule, with documentations and examples per option.

A couple of thoughts:

1. The sshKey option is a secret path, which there is no good solution for so far. See #78640
2. There's no assertion to check for presence of both `system` and `systems` at the same time, because I couldn't figure out how to do assertions per module nicely :) I guess it's an artifact anyways that there are even two options for this, I'd rather deprecate `systems`, and type `system` as str or listOf str, expecting the former to cover basically all cases.
3. As per [this suggestion](https://github.com/NixOS/nixpkgs/pull/47636#discussion_r222029830), I put the features variables an enum, and added an "extra" option that allows freeform features. The idea is to have a somewhat curated list of commonly used names where typos can't happen, and move special ones to their own option.

Ttested this locally, works as expected :)

Ericson2314 committed:
nixos/nix-daemon: Organize buildMachine options with a submodule

Valodim committed:
nixos/nix-daemon: work on buildMachines submodule

msteen commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -192,50 +192,132 @@ in
+              '';
+            };
+            system = mkOption {
+              type = types.str;
```
Should be `type = types.nullOr types.str;`.

msteen commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -192,50 +192,132 @@ in
+              '';
+            };
+            sshKey = mkOption {
+              type = types.str;
```
Should be `type = types.nullOr types.str;`.

msteen commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -192,50 +192,132 @@ in
+              description = ''
+                The relative speed of this builder. This is an arbitrary integer
+                that indicates the speed of this builder, relative to other
+                builders.
```
What does the number mean? Is higher faster? This is important to make clear, because a lot of relative numbers in Nix(OS) prioritize the lower numbers.

Valodim commented:
@msteen thanks for the review :) fixed all your points.

I also changed `system` and `systems` to default to `builtins.currentSystem` if neither are set, that seems to be a reasonable default.

msteen reviewed (approved):
Seems good to me now!

grahamc commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -192,50 +192,136 @@ in
+            supportedExtraFeatures = mkOption {
+              type = types.listOf types.str;
+              default = [];
+              example = [ "big-mistake-this-ones-impure" ];
```
Is this option impure, or you're saying there could be a feature named impure? This implies the option itself is a big mistake and impure :).

Valodim replied:
I don't know. [You tell me](https://github.com/NixOS/nixpkgs/pull/47636#discussion_r221797702) :)

grahamc replied:
This option is not impure :) there could be a feature named `impure`, though. I was trying to nudge you towards rewording the documentation through a joking hypothetical.

Infinisil commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -192,50 +192,136 @@ in
+                for mandatory custom features.
+              '';
+            };
+            mandatoryExtraFeatures = mkOption {
```
I think in general these options are rather sub-optimally laid out. E.g. currently you do
```nix
{
  buildMachines = [ {
    hostName = "example.com";
    mandatoryFeatures = [ "big-parallel" ];
  } ];
}
```

Which once declared:
- Doesn't allow you to change the config of a machine in other modules since `buildMachines` is of type `listOf`
- Even if that wasn't the case, `mandatoryFeatures` is a list still -> same problem
- In addition, due to `mandatoryFeatures` being a list, it means that one could specify the same settings multiple times, which wouldn't be a problem probably, but it's not nice

So, I'm thinking that this module could use some restructuring in the future to remove any `listOf`'s to allow overriding and such. With this expectation, I think it makes sense to currently only use this PR to make the existing module safer and more documented, such that it doesn't introduce anything other than what's there already. With this in mind, I think these `mandatoryExtraFeatures` and `supportedExtraFeatures` options can be removed, using `listOf str` for the non-Extra variant, falling in line with what it was previously.

Valodim replied:
Works for me. I'll update accordingly.

Ericson2314 commented:
@Valodim thanks for picking this up for me! And @Infinisil, thanks for letting me know.

Ericson2314 commented:
Can you add a 20.09 change log breaking change notice?

Ericson2314 commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -455,14 +514,14 @@ in
         text =
           concatMapStrings (machine:
-            "${if machine ? sshUser then "${machine.sshUser}@" else ""}${machine.hostName} "
-            + machine.system or (concatStringsSep "," machine.systems)
```
I am very against using `builtins.currentSystem` here, not just because it is impure, but because we are describing *other* systems---what the current system is is by-definition irrelevant!

Valodim replied:
Absolutely agreed! I don't know what I was thinking :) I'll fix it and bring this PR up to speed.

Ericson2314 replied:
Great! Excited to see this get merged soon :)

Valodim committed:
nixos/nix-daemon: mention potential breakage in release notes

Valodim commented:
I updated this to incorporate all feedback (and rebased on master).

I think we actually ended up compatible with (sane usage of) the previous definitions, at least the config implementation looks identical modulo handling of defaults to me, and the example from the previous documentation produces identical results.

There is a slight potential for errors in configs that used sloppy types before, e.g. a string value for `maxJobs` will lead to an error like this:
> error: The option value 'nix.buildMachines.[definition 1-entry 1].maxJobs' in ... is not of type 'signed integer'.
I added an item to the release notes about that.
